### PR TITLE
PEM BOP Functionality

### PR DESCRIPTION
**Pull request by kbrunik**
# PEM BOP Model

This allows users to include electrical balance-of-plant (BOP) in modeling the PEM electrolyzer. The electrical BOP includes a transformer and a rectifier. The model calculates a curve fit based on the provided CSV it also limits the calculated efficiencies to the maximum and minimum operating ratio values in the CSV, making the overall function a piecewise implementation.

## Impacted areas of the software
Electrical BOP is calculated at each timestep. `ECO` tools and `run_simulation()` in greenheart_simulation.py are updated to reflect changes in calculation.

Extracted some PEM electrolyzer functions into `PEM_tools.py` to remove circular import errors. 

Updated `plant_sizing_estimation.py` imports to use `PEM_tools`. Created #351 since no tests exist for this functionality.

## Test results, if applicable
Added a test for BOP functionality.

### CODE REVIEW

File: tests/greenheart/test_hydrogen/test_PEM_bop.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+from pytest import approx, fixture
```
Comment (jaredthomas68) [inline tests/greenheart/test_hydrogen/test_PEM_bop.py:1]: I'd like to see one more test with the BOP in a system to make sure it is behaving as expected. Could be a quick small system.

**Review by jaredthomas68 — CHANGES_REQUESTED** [review]
Looks pretty good. A few minor changes. Thanks!

File: greenheart/simulation/technologies/hydrogen/electrolysis/PEM_BOP/PEM_BOP.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
+import os
+import numpy as np
+import pandas as pd
+import scipy.optimize
+
+file_path = os.path.dirname(os.path.abspath(__file__))
+
+
+def calc_efficiency_curve(operating_ratio, a, b, c, d):
```
[the remark this answers is not in the document]
Comment (kbrunik) [reply greenheart/simulation/technologies/hydrogen/electrolysis/PEM_BOP/PEM_BOP.py:9]: Added reference and referred reader to markdown.

File: greenheart/tools/eco/electrolysis.py [hunk 3]
```diff
@@ -32,6 +32,8 @@
     PEM_H2_Clusters as PEMClusters,
 )
 
+from greenheart.simulation.technologies.hydrogen.electrolysis.PEM_BOP.PEM_BOP import pem_bop
```
[the remark this answers is not in the document]
Comment (kbrunik) [reply greenheart/tools/eco/electrolysis.py:35]: This one is at the top of the file but I did adjust the other one

File: greenheart/simulation/technologies/hydrogen/electrolysis/PEM_BOP/PEM_BOP.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+    Args:
+        power_profile_to_electrolyzer_kw (list or np.array): Power profile to the electrolyzer in kW.
+        electrolyzer_rated_mw (float): The rating of the PEM electrolyzer in MW.
+        electrolyzer_turn_down_ratio (float): The electrolyzer turndown ratio.
+
+    Returns:
+        energy_consumption_bop_kwh (list or np.array): Energy consumed by electrolyzer BOP in kWh.
+    """
+    from greenheart.tools.eco.electrolysis import get_electrolyzer_BOL_efficiency
```
[the remark this answers is not in the document]
Comment (kbrunik) [reply greenheart/simulation/technologies/hydrogen/electrolysis/PEM_BOP/PEM_BOP.py]: Adjusted. It was previously a circular import but I moved the tool import to the top and added a comment.

File: tests/greenheart/test_hydrogen/test_PEM_bop.py (hunk 1 above)
Comment (kbrunik) [reply tests/greenheart/test_hydrogen/test_PEM_bop.py:2]: Okay I added a small integration test for the greenheart system. Let me know if that works

File: greenheart/simulation/greenheart_simulation.py [hunk 5]
```diff
@@ -587,13 +587,13 @@ def energy_internals(
...
-                - distributed_peripheral_power
-            )
-            if r > 0:
-                remaining_power_profile_in[i] = r
+        remaining_power_profile_in = np.where(
+            hopp_results["combined_hybrid_power_production_hopp"]
+                - distributed_peripheral_power > 0,
+            hopp_results["combined_hybrid_power_production_hopp"]
+                - distributed_peripheral_power,
+            hopp_results["combined_hybrid_power_production_hopp"]
```
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply greenheart/simulation/greenheart_simulation.py]: We must not have a test sensitive to it, which is concerning.

**Review by jaredthomas68 — APPROVED** [review]
Looks good to me now

**Pull request merged**